In [1]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://postgres:postgres@localhost:5432/nyc_collisions"
)

**Q1 - How have collision counts changed by year?**

In [2]:
query = """
SELECT
    crash_year,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY crash_year
ORDER BY crash_year;
"""

q1_yearly = pd.read_sql(query, engine)

q1_yearly

,crash_year,collision_count
0,2012,85458
1,2013,171942
2,2014,172751
3,2015,182994
4,2016,192604
5,2017,217030
6,2018,216409
7,2019,194114
8,2020,104001
9,2021,101770


In [3]:
import plotly.express as px

fig = px.bar(
    q1_yearly,
    x="crash_year",
    y="collision_count",
    title="NYC Motor Vehicle Collisions by Year",
    labels={
        "crash_year": "Year",
        "collision_count": "Number of Collisions"
    }
)

fig.show()

- NYC collision frequency peaked around 2017–2018 and subsequently declined, with a particularly sharp reduction beginning in 2020.

**Q2 — Which months have the highest number of collisions?**

In [4]:
# Q2 — Collision count by month

query = """
SELECT
    crash_month,
    crash_month_name,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY
    crash_month,
    crash_month_name
ORDER BY
    crash_month;
"""

q2_monthly = pd.read_sql(query, engine)

q2_monthly

,crash_month,crash_month_name,collision_count
0,1,January,160647
1,2,February,146922
2,3,March,162271
3,4,April,146034
4,5,May,175797
5,6,June,171989
6,7,July,178157
7,8,August,176265
8,9,September,177463
9,10,October,182819


In [5]:
fig = px.bar(
    q2_monthly,
    x="crash_month_name",
    y="collision_count",
    title="NYC Motor Vehicle Collisions by Month",
    labels={
        "crash_month_name": "Month",
        "collision_count": "Number of Collisions"
    },
    text="collision_count"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Month",
    yaxis_title="Number of Collisions",
    xaxis={
        "categoryorder": "array",
        "categoryarray": [
            "January", "February", "March", "April",
            "May", "June", "July", "August",
            "September", "October", "November", "December"
        ]
    }
)

fig.show()

- Collision frequency varies by month, with October recording the highest overall collision count and April the lowest. Collision counts are generally higher from May through December than during the early months of the year.

**Q3 — What time of day has the most collisions?**

In [6]:
# Q3 — Collision count by time of day

query = """
SELECT
    time_of_day,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY time_of_day
ORDER BY collision_count DESC;
"""

q3_time_of_day = pd.read_sql(query, engine)

q3_time_of_day

,time_of_day,collision_count
0,AFTERNOON,770045
1,MORNING,528791
2,EVENING,513114
3,NIGHT,216325


In [7]:
fig = px.bar(
    q3_time_of_day,
    x="time_of_day",
    y="collision_count",
    title="NYC Collisions by Time of Day",
    labels={
        "time_of_day": "Time of Day",
        "collision_count": "Number of Collisions"
    },
    text="collision_count"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Time of Day",
    yaxis_title="Number of Collisions"
)

fig.show()

- Collisions are most frequent during the afternoon, followed by morning and evening, while nighttime has substantially fewer recorded collisions.

**Q4 — Which hours of the day have the highest collision frequency?**

In [8]:
# Q4 — Collision count by hour

query = """
SELECT
    crash_hour,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY crash_hour
ORDER BY crash_hour;
"""

q4_hourly = pd.read_sql(query, engine)

q4_hourly

,crash_hour,collision_count
0,0,68584
1,1,36344
2,2,28207
3,3,24846
4,4,28091
5,5,30253
6,6,45723
7,7,62887
8,8,110852
9,9,106104


In [9]:
fig = px.line(
    q4_hourly,
    x="crash_hour",
    y="collision_count",
    markers=True,
    title="NYC Motor Vehicle Collisions by Hour of Day",
    labels={
        "crash_hour": "Hour of Day",
        "collision_count": "Number of Collisions"
    }
)

fig.update_layout(
    xaxis_title="Hour of Day",
    yaxis_title="Number of Collisions",
    xaxis=dict(
        tickmode="linear",
        dtick=1
    )
)

fig.show()

In [14]:
# Top 5 Hours With Most Accidents

q4_hourly.sort_values(
    "collision_count",
    ascending=False
).head(5)

,crash_hour,collision_count
16,16,143488
17,17,140842
14,14,133436
15,15,126073
18,18,124201


- Collision frequency is highest during the afternoon and early evening, with 4 PM recording the highest number of collisions, followed by 5 PM. The five highest-frequency hours all fall between 2 PM and 6 PM

**Q5 — Are collisions more frequent on weekdays or weekends?**

In [11]:
# Q5 — Weekday vs Weekend collision frequency

query = """
SELECT
    CASE
        WHEN is_weekend = TRUE THEN 'Weekend'
        ELSE 'Weekday'
    END AS day_type,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY is_weekend
ORDER BY is_weekend;
"""

q5_weekday_weekend = pd.read_sql(query, engine)

q5_weekday_weekend

,day_type,collision_count
0,Weekday,1507569
1,Weekend,520706


In [12]:
fig = px.bar(
    q5_weekday_weekend,
    x="day_type",
    y="collision_count",
    title="NYC Motor Vehicle Collisions: Weekday vs Weekend",
    labels={
        "day_type": "Day Type",
        "collision_count": "Number of Collisions"
    },
    text="collision_count"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Day Type",
    yaxis_title="Number of Collisions"
)

fig.show()

- Approximately 74% of recorded collisions occurred on weekdays, compared with 26% on weekends, indicating substantially higher collision frequency during weekdays.

**Q6 — Which borough has the most collisions?**

In [15]:
# Q6 — Collision count by borough

query = """
SELECT
    borough,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
WHERE borough IS NOT NULL
GROUP BY borough
ORDER BY collision_count DESC;
"""

q6_borough = pd.read_sql(query, engine)

q6_borough

,borough,collision_count
0,BROOKLYN,496369
1,UNKNOWN,488338
2,QUEENS,413687
3,MANHATTAN,338388
4,BRONX,227119
5,STATEN ISLAND,64374


In [17]:
q6_borough["percentage"] = (
    q6_borough["collision_count"]
    / q6_borough["collision_count"].sum()
    * 100
)

q6_borough

,borough,collision_count,percentage
0,BROOKLYN,496369,24.472470
1,UNKNOWN,488338,24.076518
2,QUEENS,413687,20.396002
3,MANHATTAN,338388,16.683537
4,BRONX,227119,11.197643
5,STATEN ISLAND,64374,3.173830


In [16]:
fig = px.bar(
    q6_borough,
    x="collision_count",
    y="borough",
    orientation="h",
    title="NYC Motor Vehicle Collisions by Borough",
    labels={
        "borough": "Borough",
        "collision_count": "Number of Collisions"
    },
    text="collision_count"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Number of Collisions",
    yaxis_title="Borough",
    yaxis={
        "categoryorder": "total ascending"
    }
)

fig.show()

- Brooklyn has the highest recorded collision count among NYC's identified boroughs, with 496,369 collisions.
- A substantial number of records (488,338) lack a usable borough classification and are therefore grouped as UNKNOWN.

**Q7 — Which borough has the highest injury and fatality burden?**

In [18]:
# Q7 — Injury and fatality burden by borough

query = """
SELECT
    borough,
    COUNT(*) AS collision_count,
    SUM(number_of_persons_injured) AS total_injured,
    SUM(number_of_persons_killed) AS total_killed
FROM analytics.vw_collision_analysis
WHERE borough IS NOT NULL
GROUP BY borough
ORDER BY total_injured DESC;
"""

q7_borough_severity = pd.read_sql(query, engine)

q7_borough_severity

,borough,collision_count,total_injured,total_killed
0,UNKNOWN,488338,192481,1069
1,BROOKLYN,496369,176664,712
2,QUEENS,413687,133458,590
3,BRONX,227119,80208,313
4,MANHATTAN,338388,79073,377
5,STATEN ISLAND,64374,19649,106


In [20]:
fig_injuries = px.bar(
    q7_borough_severity,
    x="borough",
    y="total_injured",
    title="Total Persons Injured by Borough",
    labels={
        "borough": "Borough",
        "total_injured": "Persons Injured"
    },
    text="total_injured"
)

fig_injuries.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig_injuries.update_layout(
    xaxis_title="Borough",
    yaxis_title="Persons Injured"
)

fig_injuries.show()

fig_fatalities = px.bar(
    q7_borough_severity,
    x="borough",
    y="total_killed",
    title="Total Persons Killed by Borough",
    labels={
        "borough": "Borough",
        "total_killed": "Persons Killed"
    },
    text="total_killed"
)

fig_fatalities.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig_fatalities.update_layout(
    xaxis_title="Borough",
    yaxis_title="Persons Killed"
)

fig_fatalities.show()

**Key findings**
- `UNKNOWN` has the highest raw injury total (192,481) and fatality total (1,069).
- Among identified boroughs, `Brooklyn` has the highest injury burden, with `176,664 injuries`.
- `Brooklyn` also has the highest fatalities among identified boroughs, with `712 deaths`.
- `Queens` is second among identified boroughs for both 133,458 injuries and 590 fatalities
- Staten Island has the lowest totals for both measures.

**Q8 — Collision severity distribution**

In [21]:
# Q8 — Collision severity distribution

query = """
SELECT
    CASE
        WHEN number_of_persons_killed > 0 THEN 'Fatality'
        WHEN number_of_persons_injured > 0 THEN 'Injury'
        ELSE 'No Injury/Fatality'
    END AS severity_category,
    COUNT(*) AS collision_count
FROM analytics.vw_collision_analysis
GROUP BY severity_category
ORDER BY collision_count DESC;
"""

q8_severity = pd.read_sql(query, engine)

q8_severity

,severity_category,collision_count
0,No Injury/Fatality,1519677
1,Injury,505559
2,Fatality,3039


In [22]:
fig = px.bar(
    q8_severity,
    x="severity_category",
    y="collision_count",
    title="NYC Collision Severity Distribution",
    labels={
        "severity_category": "Severity Category",
        "collision_count": "Number of Collisions"
    },
    text="collision_count"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Severity Category",
    yaxis_title="Number of Collisions"
)

fig.show()

**Key findings**
- `1,519,677` collisions (≈74.9%) had no reported injury or fatality.
- `505,559` collisions (≈24.9%) involved at least one injured person.
- `3,039` collisions (≈0.15%) involved at least one fatality.
- Fatal collisions are therefore very rare compared with injury-producing collisions in the dataset.

**Q9 — Which contributing factors are most frequently associated with collisions?**

In [25]:
# Q9 — Most common contributing factors

query = """
SELECT
    f.factor_description AS contributing_factor,
    COUNT(*) AS factor_occurrences
FROM normalized.collision_factors cf
JOIN normalized.contributing_factors f
    ON cf.factor_key = f.factor_key
GROUP BY
    f.factor_description
ORDER BY
    factor_occurrences DESC;
"""

q9_factors = pd.read_sql(query, engine)

q9_factors.head(15)

,contributing_factor,factor_occurrences
0,UNSPECIFIED,2274102
1,DRIVER INATTENTION/DISTRACTION,514465
2,FAILURE TO YIELD RIGHT-OF-WAY,141456
3,FOLLOWING TOO CLOSELY,125851
4,OTHER VEHICULAR,99030
5,BACKING UNSAFELY,83445
6,PASSING OR LANE USAGE IMPROPER,72456
7,PASSING TOO CLOSELY,63626
8,TURNING IMPROPERLY,56773
9,FATIGUED/DROWSY,47305


In [24]:
q9_top15 = q9_factors.head(15).sort_values(
    "factor_occurrences",
    ascending=True
)

fig = px.bar(
    q9_top15,
    x="factor_occurrences",
    y="contributing_factor",
    orientation="h",
    title="Top 15 Contributing Factors in NYC Collisions",
    labels={
        "contributing_factor": "Contributing Factor",
        "factor_occurrences": "Factor Occurrences"
    },
    text="factor_occurrences"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Factor Occurrences",
    yaxis_title="Contributing Factor",
    height=600
)

fig.show()

- After excluding the large UNSPECIFIED category, driver inattention/distraction is the most frequently recorded contributing factor, followed by failure to yield right-of-way and following too closely. The results represent contributing-factor occurrences rather than unique collisions.

**Q10 — Which vehicle types are most frequently involved in collisions?**

In [30]:
# Q10 — Most common vehicle types involved in collisions

query = """
SELECT
    vt.vehicle_type_code,
    COUNT(*) AS vehicle_occurrences
FROM normalized.collision_vehicles cv
JOIN normalized.vehicle_types vt
    ON cv.vehicle_type_key = vt.vehicle_type_key
GROUP BY
    vt.vehicle_type_code
ORDER BY
    vehicle_occurrences DESC;
"""

q10_vehicle_types = pd.read_sql(query, engine)

q10_vehicle_types.head(15)

,vehicle_type_code,vehicle_occurrences
0,SEDAN,1094299
1,STATION WAGON/SPORT UTILITY VEHICLE,862705
2,PASSENGER VEHICLE,640442
3,SPORT UTILITY / STATION WAGON,282299
4,TAXI,148244
5,PICK-UP TRUCK,90663
6,UNKNOWN,90600
7,BUS,67874
8,VAN,62619
9,4 DR SEDAN,57967


In [29]:
q10_top15 = (
    q10_vehicle_types
    .head(15)
    .sort_values(
        "vehicle_occurrences",
        ascending=True
    )
)

fig = px.bar(
    q10_top15,
    x="vehicle_occurrences",
    y="vehicle_type_code",
    orientation="h",
    title="Top 15 Vehicle Types Involved in NYC Collisions",
    labels={
        "vehicle_type_code": "Vehicle Type",
        "vehicle_occurrences": "Vehicle Occurrences"
    },
    text="vehicle_occurrences"
)

fig.update_traces(
    texttemplate="%{text:,}",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Vehicle Occurrences",
    yaxis_title="Vehicle Type",
    height=600
)

fig.show()

**Main observations**
- `SEDAN` is by far the most frequently recorded vehicle type, with 1,094,299 occurrences.
- `STATION WAGON/SPORT UTILITY VEHICLE` is second with 862,705.
- `PASSENGER VEHICLE` is third with 640,442.
- There is a substantial gap between the first three categories and the remaining vehicle types.
- `TAXI` is the most frequent specialized/commercial category in the top 15.
- `BIKE` appears with 54,715 occurrences, while MOTORCYCLE appears with 25,156.

**Q11 — Which vehicle types are most common within each borough?**

In [31]:
# Q11 — Top vehicle types within each borough

query = """
WITH vehicle_counts AS (
    SELECT
        c.borough,
        vt.vehicle_type_code,
        COUNT(*) AS vehicle_occurrences
    FROM normalized.collisions c
    JOIN normalized.collision_vehicles cv
        ON c.collision_id = cv.collision_id
    JOIN normalized.vehicle_types vt
        ON cv.vehicle_type_key = vt.vehicle_type_key
    WHERE c.borough IS NOT NULL
      AND c.borough <> 'UNKNOWN'
      AND vt.vehicle_type_code IS NOT NULL
      AND vt.vehicle_type_code <> 'UNKNOWN'
    GROUP BY
        c.borough,
        vt.vehicle_type_code
),

ranked_vehicles AS (
    SELECT
        borough,
        vehicle_type_code,
        vehicle_occurrences,
        RANK() OVER (
            PARTITION BY borough
            ORDER BY vehicle_occurrences DESC
        ) AS vehicle_rank
    FROM vehicle_counts
)

SELECT
    borough,
    vehicle_type_code,
    vehicle_occurrences,
    vehicle_rank
FROM ranked_vehicles
WHERE vehicle_rank <= 5
ORDER BY
    borough,
    vehicle_rank;
"""

q11_borough_vehicles = pd.read_sql(query, engine)

q11_borough_vehicles

,borough,vehicle_type_code,vehicle_occurrences,vehicle_rank
0,BRONX,SEDAN,120405,1
1,BRONX,STATION WAGON/SPORT UTILITY VEHICLE,89436,2
2,BRONX,PASSENGER VEHICLE,76778,3
3,BRONX,SPORT UTILITY / STATION WAGON,28263,4
4,BRONX,TAXI,9732,5
5,BROOKLYN,SEDAN,253591,1
6,BROOKLYN,STATION WAGON/SPORT UTILITY VEHICLE,196217,2
7,BROOKLYN,PASSENGER VEHICLE,182327,3
8,BROOKLYN,SPORT UTILITY / STATION WAGON,81754,4
9,BROOKLYN,PICK-UP TRUCK,18795,5


In [32]:
fig = px.bar(
    q11_borough_vehicles,
    x="borough",
    y="vehicle_occurrences",
    color="vehicle_type_code",
    barmode="group",
    title="Top Vehicle Types by NYC Borough",
    labels={
        "borough": "Borough",
        "vehicle_occurrences": "Vehicle Occurrences",
        "vehicle_type_code": "Vehicle Type"
    }
)

fig.update_layout(
    xaxis_title="Borough",
    yaxis_title="Vehicle Occurrences",
    legend_title="Vehicle Type",
    height=600
)

fig.show()

**Borough-specific observations**
- `Brooklyn` has the highest occurrence counts across the major vehicle categories, consistent with Brooklyn having the highest overall identified-borough collision volume.
- `Queens` follows a similar vehicle composition.
- `Manhattan`is interesting because TAXI ranks 4th with 81,369 occurrences, substantially higher than in the other boroughs.
- `Bronx` has TAXI as its fifth-ranked vehicle type.
- `Brooklyn`, Queens, and Staten Island have PICK-UP TRUCK in their top five.

**Q12 — Final Question: Where are collisions concentrated?**

**Step 1 — Create the PostGIS geometry**

In [33]:
# Check PostGIS version

postgis_check = pd.read_sql(
    """
    SELECT PostGIS_Version() AS postgis_version;
    """,
    engine
)

postgis_check

,postgis_version
0,3.4 USE_GEOS=1 USE_PROJ=1 USE_STATS=1


**Cell 2 — Create spatial table**

In [34]:
query = """
DROP TABLE IF EXISTS analytics.collision_points;

CREATE TABLE analytics.collision_points AS
SELECT
    collision_id,
    crash_date,
    crash_time,
    borough,
    zip_code,
    latitude,
    longitude,
    number_of_persons_injured,
    number_of_persons_killed,

    ST_SetSRID(
        ST_MakePoint(longitude, latitude),
        4326
    ) AS geom

FROM analytics.vw_collision_analysis

WHERE latitude IS NOT NULL
  AND longitude IS NOT NULL
  AND latitude <> 0
  AND longitude <> 0
  AND latitude BETWEEN 40.5 AND 40.95
  AND longitude BETWEEN -74.25 AND -73.70;
"""

with engine.begin() as connection:
    connection.exec_driver_sql(query)

**Step 2 — Create spatial index**

In [35]:
query = """
CREATE INDEX IF NOT EXISTS idx_collision_points_geom
ON analytics.collision_points
USING GIST (geom);
"""

with engine.begin() as connection:
    connection.exec_driver_sql(query)

**Step 3 — Find collision hotspots**

In [36]:
query = """
SELECT
    latitude,
    longitude,
    COUNT(*) AS collision_count
FROM analytics.collision_points
GROUP BY
    latitude,
    longitude
ORDER BY
    collision_count DESC
LIMIT 50;
"""

q12_hotspots = pd.read_sql(query, engine)

q12_hotspots

,latitude,longitude,collision_count
0,40.608757,-74.038086,887
1,40.696033,-73.984530,712
2,40.861862,-73.912820,685
3,40.804700,-73.912430,597
4,40.696035,-73.984529,587
5,40.675735,-73.896860,581
6,40.658577,-73.890630,527
7,40.604153,-74.051980,510
8,40.798256,-73.827440,477
9,40.760601,-73.964314,474


**Step 4 — Plot the hotspots**

In [ ]:
# ! pip install folium

In [43]:
import folium
from folium.plugins import HeatMap

nyc_map = folium.Map(
    location=[40.7128, -74.0060],
    zoom_start=11
)

heat_data = q12_hotspots[
    ["latitude", "longitude", "collision_count"]
].values.tolist()

HeatMap(
    heat_data,
    radius=15,
    blur=10,
    min_opacity=0.4
).add_to(nyc_map)

nyc_map